framework/src/controls/PIDTransientControl.C

```diff
@@ -0,0 +1,148 @@
+                        "larger than the error.");
+
+  return params;
+}
+
+PIDTransientControl::PIDTransientControl(const InputParameters & parameters)
+  : Control(parameters),
+    _current(getPostprocessorValueByName(getParam<PostprocessorName>("postprocessor"))),
+    _target(getFunction("target")),
+    _Kint(getParam<Real>("K_integral")),
+    _Kpro(getParam<Real>("K_proportional")),
+    _Kder(getParam<Real>("K_derivative")),
+    _start_time(getParam<Real>("start_time")),
+    _stop_time(getParam<Real>("stop_time")),
+    _reset_every_timestep(getParam<bool>("reset_every_timestep")),
+    _reset_integral_windup(getParam<bool>("reset_integral_windup"))
+{
+  _integral = 0;
+
+  // To keep the previous values if a multiapp coupling iteration fails
+  _value_old = 0;
+  _integral_old = 0;
+
+  if (isParamValid("parameter") && isParamValid("parameter_pp"))
+    paramError("parameter_pp",
+               "Either a controllable parameter or a postprocessor to control should be specified, "
+               "not both.");
+  if (!isParamValid("parameter") && !isParamValid("parameter_pp"))
+    mooseError("A parameter or a postprocessor to control should be specified.");
+  if (isParamValid("parameter") && _reset_every_timestep)
+    paramError(
+        "reset_every_timestep",
+        "Resetting the PID every time step is only supported using controlled postprocessors");
+}
+
+void
+PIDTransientControl::execute()
+{
+  Point dummy;
+
+  if (_t >= _start_time && _t < _stop_time)
+  {
+    // Get the current value of the controllable parameter
+    Real value;
+    if (isParamValid("parameter"))
+      value = getControllableValue<Real>("parameter");
+    else
+      value = getPostprocessorValueByName(getParam<std::string>("parameter_pp"));
+
+    // Save integral and controlled value at each time step
+    // if the solver fails, a smaller time step will be used but _t_step is unchanged
+    if (_t_step != _t_step_old)
+    {
+      // Reset the error integral if PID is only used within each timestep
+      if (_reset_every_timestep)
+        _integral = 0;
+
+      _integral_old = _integral;
+      _value_old = value;
+      _t_step_old = _t_step;
+      _old_delta = 0;
+    }
+
+    // If there were coupling/Picard iterations during the transient and they failed,
+    // we need to reset the controlled value and the error integral to their initial value at the
+    // beginning of the coupling process
+    if (dynamic_cast<Transient *>(_app.getExecutioner())->picardSolve().numPicardIts() == 1)
+    {
+      _integral = _integral_old;
+      value = _value_old;
+    }
+
+    // Compute the delta between the current value of the postprocessor and the desired value
+    Real delta = _current - _target.value(_t, dummy);
+
+    // If desired, reset integral of the error if the error crosses zero
+    if (_reset_integral_windup && delta * _old_delta < 0)
+    {
+      _integral = 0;
+    }
```

Comment:
You may want to remove the `{` `}`. Looks like you do that in other places already...